
<h2 align="center">Clustering the Neighbourhoods of London and Paris</h2>

# Introduction

A Tale of Two cities, a novel written by Charles Dickens was set in London and Paris which takes place during the French Revolution. These cities were both happening then and now. A lot has changed over the years and we now take a look at how the cities have grown. 

London and Paris are quite the popular tourist and vacation destinations for people all around the world. They are diverse and multicultural and offer a wide variety of experiences that is widely sought after. We try to group the neighbourhoods of London and Paris respectively and draw insights to what they look like now.

# Business Problem


The aim is to help tourists choose their destinations depending on the experiences that the neighbourhoods have to offer and what they would want to have. This also helps people make decisions if they are thinking about migrating to London or Paris or even if they want to relocate neighbourhoods within the city. Our findings will help stakeholders make informed decisions and address any concerns they have including the different kinds of cuisines, provision stores and what the city has to offer. 


# Data Description

We require geolocation data for both London and Paris. Postal codes in each city serve as a starting point. Using Postal codes we use can find out the neighbourhoods, boroughs, venues and their most popular venue categories.


## London

To derive our solution, We scrape our data from https://en.wikipedia.org/wiki/List_of_areas_of_London

This wikipedia page has information about all the neighbourhoods, we limit it London.

1. *borough* : Name of Neighbourhood
2. *town* : Name of borough
3. *post_code* : Postal codes for London.

This wikipedia page lacks information about the geographical locations. To solve this problem we use ArcGIS API

### ArcGIS API

ArcGIS Online enables you to connect people, locations, and data using interactive maps. Work with smart, data-driven styles and intuitive analysis tools that deliver location intelligence. Share your insights with the world or specific groups. 

More specifically, we use ArcGIS to get the geo locations of the neighbourhoods of London. The following columns are added to our initial dataset which prepares our data. 

4. *latitude* : Latitude for Neighbourhood
5. *longitude* : Longitude for Neighbourhood

## Paris

To derive our solution, We leverage JSON data available at https://www.data.gouv.fr/fr/datasets/r/e88c6fda-1d09-42a0-a069-606d3259114e 

The JSON file has data about all the neighbourhoods in France, we limit it to Paris.

1. *postal_code* : Postal codes for France
2. *nom_comm* : Name of Neighbourhoods in France
3. *nom_dept* : Name of the boroughs, equivalent to towns in France
4. *geo_point_2d* : Tuple containing the latitude and longitude of the Neighbourhoods.

## Foursquare API Data

We will need data about different venues in different neighbourhoods of that specific borough. In order to gain that information we will use "Foursquare" locational information. Foursquare is a location data provider with information about all manner of venues and events within an area of interest. Such information includes venue names, locations, menus and even photos. As such, the foursquare location platform will be used as the sole data source since all the stated required information can be obtained through the API.

After finding the list of neighbourhoods, we then connect to the Foursquare API to gather information about venues inside each and every neighbourhood. For each neighbourhood, we have chosen the radius to be 500 meters.

The data retrieved from Foursquare contained information of venues within a specified distance of the longitude and latitude of the postcodes. The information obtained per venue as follows:

1. *Neighbourhood* : Name of the Neighbourhood
2. *Neighbourhood Latitude* : Latitude of the Neighbourhood
3. *Neighbourhood Longitude* : Longitude of the Neighbourhood
4. *Venue* : Name of the Venue
5. *Venue Latitude* : Latitude of Venue
6. *Venue Longitude* : Longitude of Venue
7. *Venue Category* : Category of Venue


Based on all the information collected for both London and Paris, we have sufficient data to build our model. We cluster the neighbourhoods together based on similar venue categories. We then present our observations and findings. Using this data, our stakeholders can take the necessary decision.

# Methodology

We will be creating our model with the help of Python so we start off by importing all the required packages.

In [1]:
import pandas as pd
import requests
import numpy as np
import matplotlib.cm as cm
import matplotlib.colors as colors
import folium

# import k-means for the clustering stage
from sklearn.cluster import KMeans

The approach taken here is to explore each of the cities individually, plot the map to show the neighbourhoods being considered and then build our model by clustering all of the similar neighbourhoods together and finally plot the new map with the clustered neighbourhoods. We draw insights and then compare and discuss our findings.

# Exploring London

### Neighbourhoods of London

We begin to start collecting and refining the data needed for the our business solution to work.

### Data Collection

To get the neighbourhoods in london, we start by scraping the list of areas of london wiki page.

In [4]:
url_london = "https://en.wikipedia.org/wiki/List_of_areas_of_London"
wiki_london_url = requests.get(url_london)
wiki_london_url

<Response [200]>

Response 200 means that we are able to make the connection

In [5]:
wiki_london_data = pd.read_html(wiki_london_url.text)
wiki_london_data

[                                                   0
 0  Map all coordinates in "Category:Areas of Lond...
 1                 Download coordinates as: KML · GPX,
             Location                     London borough       Post town  \
 0         Abbey Wood              Bexley, Greenwich [7]          LONDON   
 1              Acton  Ealing, Hammersmith and Fulham[8]          LONDON   
 2          Addington                         Croydon[8]         CROYDON   
 3         Addiscombe                         Croydon[8]         CROYDON   
 4        Albany Park                             Bexley  BEXLEY, SIDCUP   
 ..               ...                                ...             ...   
 527         Woolwich                          Greenwich          LONDON   
 528   Worcester Park       Sutton, Kingston upon Thames  WORCESTER PARK   
 529  Wormwood Scrubs             Hammersmith and Fulham          LONDON   
 530          Yeading                         Hillingdon           HAYES   
 

Scraping the webpage gives us all the tables present on the page. We need the 2nd table, so selecting the 2nd table.

In [6]:
wiki_london_data = wiki_london_data[1]
wiki_london_data

Location                     London borough       Post town  \
0         Abbey Wood              Bexley, Greenwich [7]          LONDON   
1              Acton  Ealing, Hammersmith and Fulham[8]          LONDON   
2          Addington                         Croydon[8]         CROYDON   
3         Addiscombe                         Croydon[8]         CROYDON   
4        Albany Park                             Bexley  BEXLEY, SIDCUP   
..               ...                                ...             ...   
527         Woolwich                          Greenwich          LONDON   
528   Worcester Park       Sutton, Kingston upon Thames  WORCESTER PARK   
529  Wormwood Scrubs             Hammersmith and Fulham          LONDON   
530          Yeading                         Hillingdon           HAYES   
531         Yiewsley                         Hillingdon    WEST DRAYTON   

    Postcode district Dial code OS grid ref  
0                 SE2       020    TQ465785  
1              W3, W4       020    TQ205805  
2                 CR0       020    TQ375645  
3                 CR0       020    TQ345665  
4           DA5, DA14       020    TQ478728  
..                ...       ...         ...  
527              SE18       020    TQ435795  
528               KT4       020    TQ225655  
529               W12       020    TQ225815  
530               UB4       020    TQ115825  
531               UB7       020    TQ063804  

[532 rows x 6 columns]

### Data Preprocessing

we remove the spaces in the column titles and then we add _ between words.

In [7]:
wiki_london_data.rename(columns=lambda x: x.strip().replace(" ", "_"), inplace=True)
wiki_london_data

Location                     London borough       Post_town  \
0         Abbey Wood              Bexley, Greenwich [7]          LONDON   
1              Acton  Ealing, Hammersmith and Fulham[8]          LONDON   
2          Addington                         Croydon[8]         CROYDON   
3         Addiscombe                         Croydon[8]         CROYDON   
4        Albany Park                             Bexley  BEXLEY, SIDCUP   
..               ...                                ...             ...   
527         Woolwich                          Greenwich          LONDON   
528   Worcester Park       Sutton, Kingston upon Thames  WORCESTER PARK   
529  Wormwood Scrubs             Hammersmith and Fulham          LONDON   
530          Yeading                         Hillingdon           HAYES   
531         Yiewsley                         Hillingdon    WEST DRAYTON   

    Postcode district Dial code OS_grid_ref  
0                 SE2       020    TQ465785  
1              W3, W4       020    TQ205805  
2                 CR0       020    TQ375645  
3                 CR0       020    TQ345665  
4           DA5, DA14       020    TQ478728  
..                ...       ...         ...  
527              SE18       020    TQ435795  
528               KT4       020    TQ225655  
529               W12       020    TQ225815  
530               UB4       020    TQ115825  
531               UB7       020    TQ063804  

[532 rows x 6 columns]

We see that few columns have no '_' between the words despite applying our function meaning that there are special characters

### Feature Selection

We need only the boroughs, Postal codes, Post town for further steps. We can drop the locations, dial codes and OS grid.

In [8]:
df1 = wiki_london_data.drop( [ wiki_london_data.columns[0], wiki_london_data.columns[4], wiki_london_data.columns[5] ], axis=1)

In [9]:
df1.head()

London borough       Post_town Postcode district
0              Bexley, Greenwich [7]          LONDON               SE2
1  Ealing, Hammersmith and Fulham[8]          LONDON            W3, W4
2                         Croydon[8]         CROYDON               CR0
3                         Croydon[8]         CROYDON               CR0
4                             Bexley  BEXLEY, SIDCUP         DA5, DA14

let's rename the Postcode district column and the london borough to something simpler

In [10]:
df1.columns = ['borough','town','post_code']
df1

borough            town  post_code
0                Bexley, Greenwich [7]          LONDON        SE2
1    Ealing, Hammersmith and Fulham[8]          LONDON     W3, W4
2                           Croydon[8]         CROYDON        CR0
3                           Croydon[8]         CROYDON        CR0
4                               Bexley  BEXLEY, SIDCUP  DA5, DA14
..                                 ...             ...        ...
527                          Greenwich          LONDON       SE18
528       Sutton, Kingston upon Thames  WORCESTER PARK        KT4
529             Hammersmith and Fulham          LONDON        W12
530                         Hillingdon           HAYES        UB4
531                         Hillingdon    WEST DRAYTON        UB7

[532 rows x 3 columns]

Let's remove the Square brackets [ ] and numbers from the borough column

In [11]:
df1['borough'] = df1['borough'].map(lambda x: x.rstrip(']').rstrip('0123456789').rstrip('['))
df1

borough            town  post_code
0                Bexley, Greenwich           LONDON        SE2
1    Ealing, Hammersmith and Fulham          LONDON     W3, W4
2                           Croydon         CROYDON        CR0
3                           Croydon         CROYDON        CR0
4                            Bexley  BEXLEY, SIDCUP  DA5, DA14
..                              ...             ...        ...
527                       Greenwich          LONDON       SE18
528    Sutton, Kingston upon Thames  WORCESTER PARK        KT4
529          Hammersmith and Fulham          LONDON        W12
530                      Hillingdon           HAYES        UB4
531                      Hillingdon    WEST DRAYTON        UB7

[532 rows x 3 columns]

Take the dimension of the dataframe

In [12]:
df1.shape

(532, 3)

We currently have 533 records and 3 columns of our data. It's time to perform Feature Engineering

### Feature Engineering

We can only focusing on the neighbourhoods of London, so performing the changes

In [13]:
df1 = df1[df1['town'].str.contains('LONDON')]
df1

borough                    town post_code
0                Bexley, Greenwich                   LONDON       SE2
1    Ealing, Hammersmith and Fulham                  LONDON    W3, W4
6                              City                  LONDON       EC3
7                       Westminster                  LONDON       WC2
9                           Bromley                  LONDON      SE20
..                              ...                     ...       ...
522                       Redbridge                  LONDON  IG8, E18
523       Redbridge, Waltham Forest  LONDON, WOODFORD GREEN       IG8
526                          Barnet                  LONDON       N12
527                       Greenwich                  LONDON      SE18
529          Hammersmith and Fulham                  LONDON       W12

[309 rows x 3 columns]

In [14]:
df1.shape

(309, 3)

We now have only 310 rows. We can proceed with our further steps. Getting some descriptive statistics

In [15]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 309 entries, 0 to 529
Data columns (total 3 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   borough    309 non-null    object
 1   town       309 non-null    object
 2   post_code  309 non-null    object
dtypes: object(3)
memory usage: 9.7+ KB


## Geolocations of the London Neighbourhoods

### ArcGis API

We need to get the geographical co-ordinates for the neighbourhoods to plot out map. We will use the arcgis package to do so. 

Arcgis doesn't have a limitation on the number of API calls made so it fits our use case perfectly.

In [16]:
pip install arcgis

Processing c:\users\danikusumah\appdata\local\pip\cache\wheels\4d\11\58\7d0a04db6c902ef42b717da2981807529f4922485141ab404f\lerc-0.1.0-py3-none-any.whl
  Using cached python_certifi_win32-1.6-py2.py3-none-any.whl (7.2 kB)
Processing c:\users\danikusumah\appdata\local\pip\cache\wheels\49\29\69\6b6a800e28c596146549d22f4e1013abfcbb466cd179e586d3\pyshp-2.1.2-py3-none-any.whl
  Using cached requests_oauthlib-1.3.0-py2.py3-none-any.whl (23 kB)
  Using cached requests_toolbelt-0.9.1-py2.py3-none-any.whl (54 kB)
  Using cached requests_ntlm-1.1.0-py2.py3-none-any.whl (5.7 kB)
  Using cached requests_negotiate_sspi-0.5.2-py2.py3-none-any.whl (7.1 kB)
  Using cached requests_kerberos-0.12.0-py2.py3-none-any.whl (14 kB)
  Using cached winkerberos-0.7.0-cp38-cp38-win_amd64.whl (23 kB)
Note: you may need to restart the kernel to use updated packages.Requirement already satisfied: notebook>=4.4.1 in c:\programdata\anaconda3\lib\site-packages (from widgetsnbextension>=3->arcgis) (6.1.4)

  Using cache

In [17]:
from arcgis.geocoding import geocode
from arcgis.gis import GIS
gis = GIS()

Defining London arcgis geocode function to return latitude and longitude

In [18]:
def get_x_y_uk(address1):
   lat_coords = 0
   lng_coords = 0
   g = geocode(address='{}, London, England, GBR'.format(address1))[0]
   lng_coords = g['location']['x']
   lat_coords = g['location']['y']
   return str(lat_coords) +","+ str(lng_coords)

Checking sample data

In [19]:
c = get_x_y_uk('SE2')

In [18]:
c

'51.492450000000076,0.12127000000003818'

Looks good, We Copy over the postal codes of london to pass it into the geolocator function that we just defined above

In [20]:
geo_coordinates_uk = df1['post_code']    
geo_coordinates_uk

0           SE2
1        W3, W4
6           EC3
7           WC2
9          SE20
         ...   
522    IG8, E18
523         IG8
526         N12
527        SE18
529         W12
Name: post_code, Length: 309, dtype: object

Passing postal codes of london to get the geographical co-ordinates

In [21]:
coordinates_latlng_uk = geo_coordinates_uk.apply(lambda x: get_x_y_uk(x))
coordinates_latlng_uk

0       51.492450000000076,0.12127000000003818
1        51.51324000000005,-0.2674599999999714
6       51.51200000000006,-0.08057999999994081
7       51.51651000000004,-0.11967999999995982
9       51.41009000000008,-0.05682999999993399
                        ...                   
522    51.589770000000044,0.030520000000024083
523      51.50642000000005,-0.1272099999999341
526     51.615920000000074,-0.1767399999999384
527      51.48207000000008,0.07143000000002075
529      51.50645000000003,-0.2369099999999662
Name: post_code, Length: 309, dtype: object

### Latitude

Extracting the latitude from our previously collected coordinates

In [22]:
lat_uk = coordinates_latlng_uk.apply(lambda x: x.split(',')[0])
lat_uk

0      51.492450000000076
1       51.51324000000005
6       51.51200000000006
7       51.51651000000004
9       51.41009000000008
              ...        
522    51.589770000000044
523     51.50642000000005
526    51.615920000000074
527     51.48207000000008
529     51.50645000000003
Name: post_code, Length: 309, dtype: object

### Longitude

Extracting the Longitude from our previously collected coordinates

In [23]:
lng_uk = coordinates_latlng_uk.apply(lambda x: x.split(',')[1])
lng_uk

0       0.12127000000003818
1       -0.2674599999999714
6      -0.08057999999994081
7      -0.11967999999995982
9      -0.05682999999993399
               ...         
522    0.030520000000024083
523     -0.1272099999999341
526     -0.1767399999999384
527     0.07143000000002075
529     -0.2369099999999662
Name: post_code, Length: 309, dtype: object

We now have the geographical co-ordinates of the London Neighbourhoods.

We proceed with Merging our source data with the geographical co-ordinates to make our dataset ready for the next stage

In [24]:
london_merged = pd.concat([df1,lat_uk.astype(float), lng_uk.astype(float)], axis=1)
london_merged.columns= ['borough','town','post_code','latitude','longitude']
london_merged

borough                    town post_code  \
0                Bexley, Greenwich                   LONDON       SE2   
1    Ealing, Hammersmith and Fulham                  LONDON    W3, W4   
6                              City                  LONDON       EC3   
7                       Westminster                  LONDON       WC2   
9                           Bromley                  LONDON      SE20   
..                              ...                     ...       ...   
522                       Redbridge                  LONDON  IG8, E18   
523       Redbridge, Waltham Forest  LONDON, WOODFORD GREEN       IG8   
526                          Barnet                  LONDON       N12   
527                       Greenwich                  LONDON      SE18   
529          Hammersmith and Fulham                  LONDON       W12   

     latitude  longitude  
0    51.49245    0.12127  
1    51.51324   -0.26746  
6    51.51200   -0.08058  
7    51.51651   -0.11968  
9    51.41009   -0.05683  
..        ...        ...  
522  51.58977    0.03052  
523  51.50642   -0.12721  
526  51.61592   -0.17674  
527  51.48207    0.07143  
529  51.50645   -0.23691  

[309 rows x 5 columns]

In [25]:
london_merged.dtypes

borough       object
town          object
post_code     object
latitude     float64
longitude    float64
dtype: object

### Co-ordinates for London

Getting the geocode for London to help visualize it on the map

In [26]:
london = geocode(address='London, England, GBR')[0]
london_lng_coords = london['location']['x']
london_lat_coords = london['location']['y']
london_lng_coords

-0.1272099999999341

In [27]:
london_lat_coords

51.50642000000005

## Visualize the Map of London

To help visualize the Map of London and the neighbourhoods in London, we make use of the folium package.

In [28]:
# Creating the map of London
map_London = folium.Map(location=[london_lat_coords, london_lng_coords], zoom_start=12)
map_London

# adding markers to map
for latitude, longitude, borough, town in zip(london_merged['latitude'], london_merged['longitude'], london_merged['borough'], london_merged['town']):
    label = '{}, {}'.format(town, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='red',
        fill=True
        ).add_to(map_London)  
    
map_London

### Venues in London

To proceed with the next part, we need to define Foursquare API credentials.

Using Foursquare API, we are able to get the venue and venue categories around each neighbourhood in London.

In [29]:
CLIENT_ID = 'MCVLEZYYF4O5SBWB5XH1BJIIJD4AOE4FPLA3LLKAJWKIYOTW' 
CLIENT_SECRET = 'HQEBAIHA5LJH5KBLSFVZZVV2BXYCU22CTNJWXMDNV2R5RXLS'
VERSION = '20180605' # Foursquare API version

Defining a function to get the neraby venues in the neighbourhood. This will help us get venue categories which is important for our analysis

In [31]:
LIMIT=100

def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius,
            LIMIT
            )
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Category']
    
    return(nearby_venues)

Getting the venues in London

In [32]:
venues_in_London = getNearbyVenues(london_merged['borough'], london_merged['latitude'], london_merged['longitude'])

Bexley, Greenwich 
Ealing, Hammersmith and Fulham
City
Westminster
Bromley
Islington
Islington
Barnet
Enfield
Wandsworth
Southwark
City
Richmond upon Thames
Barnet
Islington
Wandsworth
Westminster
Bromley
Newham
Ealing
Westminster
Lewisham
Camden
Southwark
Tower Hamlets
Bexley
City
Lewisham
Greenwich
Tower Hamlets
Camden
Haringey
Tower Hamlets
Haringey
Barnet
Brent
Lambeth
Lewisham
Tower Hamlets
Kensington and Chelsea, Hammersmith and Fulham
Brent
Barnet
Barnet
Southwark
Tower Hamlets
Camden
Tower Hamlets
Waltham Forest
Newham
Islington
Richmond upon Thames
Lewisham
Camden
Westminster
Greenwich
Kensington and Chelsea
Barnet
Westminster
Lewisham
Waltham Forest
Hounslow, Ealing, Hammersmith and Fulham
Brent
Barnet
Lambeth, Wandsworth
Islington
Barnet
Merton
Barnet
Westminster
Barnet, Brent, Camden
Lewisham
Bexley
Haringey
Bromley
Tower Hamlets
Newham
Hackney
Islington
Southwark
Lewisham
Brent
Southwark
Ealing
Kensington and Chelsea
Wandsworth
Southwark
Barnet
Newham
Richmond upon Thames


Sampling our data

In [33]:
venues_in_London.head()

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0  Bexley, Greenwich                 51.49245                  0.12127   
1  Bexley, Greenwich                 51.49245                  0.12127   
2  Bexley, Greenwich                 51.49245                  0.12127   
3  Bexley, Greenwich                 51.49245                  0.12127   
4  Bexley, Greenwich                 51.49245                  0.12127   

                              Venue Venue Category  
0                      Lesnes Abbey  Historic Site  
1                       Sainsbury's    Supermarket  
2                              Lidl    Supermarket  
3  Abbey Wood Railway Station (ABW)  Train Station  
4                       Bean @ Work    Coffee Shop

In [34]:
venues_in_London.shape

(10501, 5)

Wow, we have scraped together 10567 records for venues. This will definitely make the clustering interesting.



### Grouping by Venue Categories
We need to now see how many Venue Categories are there for further processing

In [35]:
venues_in_London.groupby('Venue Category').max()

Neighbourhood  \
Venue Category                                                        
Accessories Store                                       Westminster   
Adult Boutique                                            Islington   
African Restaurant                                      Westminster   
American Restaurant                                  Waltham Forest   
Antique Shop                                            Westminster   
...                                                             ...   
Wings Joint                                  Hammersmith and Fulham   
Women's Store        Kensington and Chelsea, Hammersmith and Fulham   
Xinjiang Restaurant                                       Southwark   
Yoga Studio                                             Westminster   
Zoo Exhibit                                                  Camden   

                     Neighbourhood Latitude  Neighbourhood Longitude  \
Venue Category                                                         
Accessories Store                  51.51651                 -0.11968   
Adult Boutique                     51.52969                 -0.08697   
African Restaurant                 51.52587                 -0.08808   
American Restaurant                51.61780                  0.02795   
Antique Shop                       51.51651                 -0.11968   
...                                     ...                      ...   
Wings Joint                        51.54187                 -0.19795   
Women's Store                      51.55457                 -0.11478   
Xinjiang Restaurant                51.47480                 -0.09313   
Yoga Studio                        51.55457                 -0.03558   
Zoo Exhibit                        51.53354                 -0.14606   

                                           Venue  
Venue Category                                    
Accessories Store             James Smith & Sons  
Adult Boutique       Sh! Women's Erotic Emporium  
African Restaurant            Red Sea Restaurant  
American Restaurant                  Spielburger  
Antique Shop            The London Silver Vaults  
...                                          ...  
Wings Joint                             Wingmans  
Women's Store                 Vivien of Holloway  
Xinjiang Restaurant                    Silk Road  
Yoga Studio                            yogahaven  
Zoo Exhibit                        Penguin Beach  

[301 rows x 4 columns]

We can see 295 records, just goes to show how diverse and interesting the place is.

### One Hot Encoding 
We need to Encode our venue categories to get a better result for our clustering

In [36]:
London_venue_cat = pd.get_dummies(venues_in_London[['Venue Category']], prefix="", prefix_sep="")
London_venue_cat

Accessories Store  Adult Boutique  African Restaurant  \
0                      0               0                   0   
1                      0               0                   0   
2                      0               0                   0   
3                      0               0                   0   
4                      0               0                   0   
...                  ...             ...                 ...   
10496                  0               0                   0   
10497                  0               0                   0   
10498                  0               0                   0   
10499                  0               0                   0   
10500                  0               0                   0   

       American Restaurant  Antique Shop  Arcade  Arepa Restaurant  \
0                        0             0       0                 0   
1                        0             0       0                 0   
2                        0             0       0                 0   
3                        0             0       0                 0   
4                        0             0       0                 0   
...                    ...           ...     ...               ...   
10496                    0             0       0                 0   
10497                    0             0       0                 0   
10498                    0             0       0                 0   
10499                    0             0       0                 0   
10500                    0             0       0                 0   

       Argentinian Restaurant  Art Gallery  Art Museum  ...  Warehouse Store  \
0                           0            0           0  ...                0   
1                           0            0           0  ...                0   
2                           0            0           0  ...                0   
3                           0            0           0  ...                0   
4                           0            0           0  ...                0   
...                       ...          ...         ...  ...              ...   
10496                       0            0           0  ...                0   
10497                       0            0           0  ...                0   
10498                       0            0           0  ...                0   
10499                       0            0           0  ...                0   
10500                       0            0           0  ...                0   

       Watch Shop  Whisky Bar  Wine Bar  Wine Shop  Wings Joint  \
0               0           0         0          0            0   
1               0           0         0          0            0   
2               0           0         0          0            0   
3               0           0         0          0            0   
4               0           0         0          0            0   
...           ...         ...       ...        ...          ...   
10496           0           0         0          0            0   
10497           0           0         0          0            0   
10498           0           0         0          0            0   
10499           0           0         0          0            0   
10500           0           0         0          0            0   

       Women's Store  Xinjiang Restaurant  Yoga Studio  Zoo Exhibit  
0                  0                    0            0            0  
1                  0                    0            0            0  
2                  0                    0            0            0  
3                  0                    0            0            0  
4                  0                    0            0            0  
...              ...                  ...          ...          ...  
10496              0                    0            0            0  
10497              0                    0            0            0  
1

Adding Neighbourhood into the mix.

In [37]:
London_venue_cat['Neighbourhood'] = venues_in_London['Neighbourhood'] 

# moving neighborhood column to the first column
fixed_columns = [London_venue_cat.columns[-1]] + list(London_venue_cat.columns[:-1])
London_venue_cat = London_venue_cat[fixed_columns]

London_venue_cat.head()

Neighbourhood  Accessories Store  Adult Boutique  African Restaurant  \
0  Bexley, Greenwich                   0               0                   0   
1  Bexley, Greenwich                   0               0                   0   
2  Bexley, Greenwich                   0               0                   0   
3  Bexley, Greenwich                   0               0                   0   
4  Bexley, Greenwich                   0               0                   0   

   American Restaurant  Antique Shop  Arcade  Arepa Restaurant  \
0                    0             0       0                 0   
1                    0             0       0                 0   
2                    0             0       0                 0   
3                    0             0       0                 0   
4                    0             0       0                 0   

   Argentinian Restaurant  Art Gallery  ...  Warehouse Store  Watch Shop  \
0                       0            0  ...                0           0   
1                       0            0  ...                0           0   
2                       0            0  ...                0           0   
3                       0            0  ...                0           0   
4                       0            0  ...                0           0   

   Whisky Bar  Wine Bar  Wine Shop  Wings Joint  Women's Store  \
0           0         0          0            0              0   
1           0         0          0            0              0   
2           0         0          0            0              0   
3           0         0          0            0              0   
4           0         0          0            0              0   

   Xinjiang Restaurant  Yoga Studio  Zoo Exhibit  
0                    0            0            0  
1                    0            0            0  
2                    0            0            0  
3                    0            0            0  
4                    0            0            0  

[5 rows x 302 columns]

### Venue categories mean value
We will group the Neighbourhoods and calculate the mean venue categories value in each Neighbourhood

In [38]:
London_grouped = London_venue_cat.groupby('Neighbourhood').mean().reset_index()
London_grouped.head()

Neighbourhood  Accessories Store  Adult Boutique  \
0                 Barnet                0.0             0.0   
1  Barnet, Brent, Camden                0.0             0.0   
2                 Bexley                0.0             0.0   
3      Bexley, Greenwich                0.0             0.0   
4     Bexley, Greenwich                 0.0             0.0   

   African Restaurant  American Restaurant  Antique Shop  Arcade  \
0                 0.0             0.001832           0.0     0.0   
1                 0.0             0.000000           0.0     0.0   
2                 0.0             0.000000           0.0     0.0   
3                 0.0             0.000000           0.0     0.0   
4                 0.0             0.000000           0.0     0.0   

   Arepa Restaurant  Argentinian Restaurant  Art Gallery  ...  \
0               0.0                0.007326          0.0  ...   
1               0.0                0.000000          0.0  ...   
2               0.0                0.000000          0.0  ...   
3               0.0                0.000000          0.0  ...   
4               0.0                0.000000          0.0  ...   

   Warehouse Store  Watch Shop  Whisky Bar  Wine Bar  Wine Shop  Wings Joint  \
0              0.0         0.0         0.0       0.0        0.0          0.0   
1              0.0         0.0         0.0       0.0        0.0          0.0   
2              0.0         0.0         0.0       0.0        0.0          0.0   
3              0.0         0.0         0.0       0.0        0.0          0.0   
4              0.0         0.0         0.0       0.0        0.0          0.0   

   Women's Store  Xinjiang Restaurant  Yoga Studio  Zoo Exhibit  
0            0.0                  0.0          0.0          0.0  
1            0.0                  0.0          0.0          0.0  
2            0.0                  0.0          0.0          0.0  
3            0.0                  0.0          0.0          0.0  
4            0.0                  0.0          0.0          0.0  

[5 rows x 302 columns]


Let's make a function to get the top most common venue categories

In [39]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]


There are way too many venue categories, we can take the top 10 to cluster the neighbourhoods.

Creating a function to label the columns of the venue correctly

In [40]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))


### Top venue categories

Getting the top venue categories in London

In [41]:
# create a new dataframe for London
neighborhoods_venues_sorted_london = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted_london['Neighbourhood'] = London_grouped['Neighbourhood']

for ind in np.arange(London_grouped.shape[0]):
    neighborhoods_venues_sorted_london.iloc[ind, 1:] = return_most_common_venues(London_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted_london.head()

Neighbourhood 1st Most Common Venue 2nd Most Common Venue  \
0                 Barnet           Coffee Shop                  Café   
1  Barnet, Brent, Camden  Gym / Fitness Center              Bus Stop   
2                 Bexley           Supermarket         Historic Site   
3      Bexley, Greenwich                  Park              Bus Stop   
4     Bexley, Greenwich            Supermarket           Coffee Shop   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0         Grocery Store                   Pub    Italian Restaurant   
1        Clothing Store           Supermarket        Hardware Store   
2         Train Station           Coffee Shop              Platform   
3        Massage Studio     Convenience Store          Home Service   
4     Convenience Store              Platform         Train Station   

        6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0                    Pharmacy           Supermarket              Bus Stop   
1                 Zoo Exhibit   Filipino Restaurant               Exhibit   
2           Convenience Store                  Park           Golf Course   
3  Construction & Landscaping         Historic Site           Golf Course   
4               Historic Site           Event Space               Exhibit   

        9th Most Common Venue 10th Most Common Venue  
0            Sushi Restaurant     Turkish Restaurant  
1          Falafel Restaurant         Farmers Market  
2  Construction & Landscaping               Bus Stop  
3        Fast Food Restaurant   Ethiopian Restaurant  
4          Falafel Restaurant         Farmers Market

## Model Building

### K Means
Let's cluster the city of london to roughly 5 to make it easier to analyze. 

We use the K Means clustering technique to do so.

In [42]:
# set number of clusters
k_num_clusters = 5

London_grouped_clustering = London_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans_london = KMeans(n_clusters=k_num_clusters, random_state=0).fit(London_grouped_clustering)
kmeans_london

KMeans(n_clusters=5, random_state=0)

### Labelling Clustered Data

In [43]:
kmeans_london.labels_

array([3, 4, 0, 2, 0, 3, 3, 3, 3, 0, 3, 3, 3, 3, 3, 3, 1, 3, 3, 3, 3, 3,
       3, 3, 3, 2, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 2, 3, 3, 3, 3,
       3, 3, 3, 3, 3, 3])

So our model has labeled the city

In [44]:
neighborhoods_venues_sorted_london.insert(0, 'Cluster Labels', kmeans_london.labels_ +1)

Join London_merged with our neighbourhood venues sorted to add latitude & longitude for each of the neighborhood to prepare it for plotting

In [45]:
london_data = london_merged

london_data = london_data.join(neighborhoods_venues_sorted_london.set_index('Neighbourhood'), on='borough')

london_data.head()

borough    town post_code  latitude  longitude  \
0              Bexley, Greenwich   LONDON       SE2  51.49245    0.12127   
1  Ealing, Hammersmith and Fulham  LONDON    W3, W4  51.51324   -0.26746   
6                            City  LONDON       EC3  51.51200   -0.08058   
7                     Westminster  LONDON       WC2  51.51651   -0.11968   
9                         Bromley  LONDON      SE20  51.41009   -0.05683   

   Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0               1           Supermarket           Coffee Shop   
1               2         Grocery Store     Indian Restaurant   
6               4                 Hotel           Coffee Shop   
7               4                 Hotel           Coffee Shop   
9               1           Supermarket         Grocery Store   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0     Convenience Store              Platform         Train Station   
1         Train Station                  Park        Breakfast Spot   
6    Italian Restaurant  Gym / Fitness Center                   Pub   
7                   Pub                  Café    Italian Restaurant   
9     Convenience Store                 Hotel  Fast Food Restaurant   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0         Historic Site           Event Space               Exhibit   
1                 Hotel                Hostel  Ethiopian Restaurant   
6        Sandwich Place            Restaurant        Scenic Lookout   
7        Sandwich Place               Theater            Restaurant   
9                  Park                   Pub              Bus Stop   

  9th Most Common Venue 10th Most Common Venue  
0    Falafel Restaurant         Farmers Market  
1           Event Space                Exhibit  
6    Falafel Restaurant      French Restaurant  
7     French Restaurant           Burger Joint  
9                  Café     Chinese Restaurant


Drop all the NaN values to prevent data skew

In [46]:
london_data_nonan = london_data.dropna(subset=['Cluster Labels'])

### Visualizing the clustered neighbourhood
Let's plot the clusters

In [47]:
map_clusters_london = folium.Map(location=[london_lat_coords, london_lng_coords], zoom_start=12)

# set color scheme for the clusters
x = np.arange(k_num_clusters)
ys = [i + x + (i*x)**2 for i in range(k_num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(london_data_nonan['latitude'], london_data_nonan['longitude'], london_data_nonan['borough'], london_data_nonan['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + '\n' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)]
        ).add_to(map_clusters_london)
        
map_clusters_london

## Examining our Clusters

Cluster 1

In [48]:
london_data_nonan.loc[london_data_nonan['Cluster Labels'] == 1, london_data_nonan.columns[[1] + list(range(5, london_data_nonan.shape[1]))]]

town  Cluster Labels 1st Most Common Venue  \
0                 LONDON               1           Supermarket   
9                 LONDON               1           Supermarket   
29     BECKENHAM, LONDON               1           Supermarket   
45   BEXLEYHEATH, LONDON               1           Supermarket   
124               LONDON               1           Supermarket   
127               LONDON               1           Supermarket   
291       LONDON, SIDCUP               1           Supermarket   
316               LONDON               1           Supermarket   
360               LONDON               1           Supermarket   
506               LONDON               1           Supermarket   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0             Coffee Shop     Convenience Store              Platform   
9           Grocery Store     Convenience Store                 Hotel   
29          Grocery Store     Convenience Store                 Hotel   
45          Historic Site         Train Station           Coffee Shop   
124         Historic Site         Train Station           Coffee Shop   
127         Grocery Store     Convenience Store                 Hotel   
291         Historic Site         Train Station           Coffee Shop   
316         Grocery Store     Convenience Store                 Hotel   
360         Grocery Store     Convenience Store                 Hotel   
506         Historic Site         Train Station           Coffee Shop   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0           Train Station         Historic Site           Event Space   
9    Fast Food Restaurant                  Park                   Pub   
29   Fast Food Restaurant                  Park                   Pub   
45               Platform     Convenience Store                  Park   
124              Platform     Convenience Store                  Park   
127  Fast Food Restaurant                  Park                   Pub   
291              Platform     Convenience Store                  Park   
316  Fast Food Restaurant                  Park                   Pub   
360  Fast Food Restaurant                  Park                   Pub   
506              Platform     Convenience Store                  Park   

    8th Most Common Venue       9th Most Common Venue 10th Most Common Venue  
0                 Exhibit          Falafel Restaurant         Farmers Market  
9                Bus Stop                        Café     Chinese Restaurant  
29               Bus Stop                        Café     Chinese Restaurant  
45            Golf Course  Construction & Landscaping               Bus Stop  
124           Golf Course  Construction & Landscaping               Bus Stop  
127              Bus Stop                        Café     Chinese Restaurant  
291           Golf Course  Construction & Landscaping               Bus Stop  
316              Bus Stop                        Café     Chinese Restaurant  
360              Bus Stop                        Café     Chinese Restaurant  
506           Golf Course  Construction & Landscaping               Bus Stop

Cluster 2

In [49]:
london_data_nonan.loc[london_data_nonan['Cluster Labels'] == 2, london_data_nonan.columns[[1] + list(range(5, london_data_nonan.shape[1]))]]

town  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
1  LONDON               2         Grocery Store     Indian Restaurant   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
1         Train Station                  Park        Breakfast Spot   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
1                 Hotel                Hostel  Ethiopian Restaurant   

  9th Most Common Venue 10th Most Common Venue  
1           Event Space                Exhibit

Cluster 3

In [50]:
london_data_nonan.loc[london_data_nonan['Cluster Labels'] == 3, london_data_nonan.columns[[1] + list(range(5, london_data_nonan.shape[1]))]]

town  Cluster Labels 1st Most Common Venue  \
167                  LONDON, WELLING               3                  Park   
378  HARROW, STANMOREEDGWARE, LONDON               3                Bakery   
454                           LONDON               3                  Park   
458                    LONDON, ERITH               3                  Park   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
167              Bus Stop        Massage Studio     Convenience Store   
378                   Gym                  Park              Bus Stop   
454            Restaurant           Flower Shop                   Pub   
458              Bus Stop        Massage Studio     Convenience Store   

    5th Most Common Venue       6th Most Common Venue 7th Most Common Venue  \
167          Home Service  Construction & Landscaping         Historic Site   
378         Metro Station                 Film Studio               Exhibit   
454             Wine Shop                Tennis Court         Train Station   
458          Home Service  Construction & Landscaping         Historic Site   

    8th Most Common Venue  9th Most Common Venue 10th Most Common Venue  
167           Golf Course   Fast Food Restaurant   Ethiopian Restaurant  
378    Falafel Restaurant         Farmers Market   Fast Food Restaurant  
454  Gym / Fitness Center  Entertainment Service            Escape Room  
458           Golf Course   Fast Food Restaurant   Ethiopian Restaurant

Cluster 4

In [51]:
london_data_nonan.loc[london_data_nonan['Cluster Labels'] == 4, london_data_nonan.columns[[1] + list(range(5, london_data_nonan.shape[1]))]]

town  Cluster Labels 1st Most Common Venue  \
6                    LONDON               4                 Hotel   
7                    LONDON               4                 Hotel   
10                   LONDON               4                   Pub   
12                   LONDON               4                   Pub   
14           BARNET, LONDON               4           Coffee Shop   
..                      ...             ...                   ...   
522                  LONDON               4         Grocery Store   
523  LONDON, WOODFORD GREEN               4               Theater   
526                  LONDON               4           Coffee Shop   
527                  LONDON               4                   Pub   
529                  LONDON               4                   Pub   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
6             Coffee Shop    Italian Restaurant  Gym / Fitness Center   
7             Coffee Shop                   Pub                  Café   
10            Coffee Shop                  Café            Food Truck   
12            Coffee Shop                  Café            Food Truck   
14                   Café         Grocery Store                   Pub   
..                    ...                   ...                   ...   
522                   Pub                   Bar                  Café   
523                 Hotel                 Plaza                Garden   
526                  Café         Grocery Store                   Pub   
527         Grocery Store              Bus Stop           Coffee Shop   
529                  Café           Coffee Shop         Grocery Store   

     5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
6                      Pub        Sandwich Place            Restaurant   
7       Italian Restaurant        Sandwich Place               Theater   
10   Vietnamese Restaurant  Gym / Fitness Center                  Park   
12   Vietnamese Restaurant  Gym / Fitness Center                  Park   
14      Italian Restaurant              Pharmacy           Supermarket   
..                     ...                   ...                   ...   
522     Seafood Restaurant             BBQ Joint    Italian Restaurant   
523                   Café                   Pub            Restaurant   
526     Italian Restaurant              Pharmacy           Supermarket   
527      Indian Restaurant                  Pier    Turkish Restaurant   
529              Gastropub                 Hotel       Thai Restaurant   

    8th Most Common Venue       9th Most Common Venue 10th Most Common Venue  
6          Scenic Lookout          Falafel Restaurant      French Restaurant  
7              Restaurant           French Restaurant           Burger Joint  
10         Breakfast Spot                Cocktail Bar                  Hotel  
12         Breakfast Spot                Cocktail Bar                  Hotel  
14               Bus Stop            Sushi Restaurant     Turkish Restaurant  
..                    ...                         ...                    ...  
522                Bakery                 Coffee Shop      Indian Restaurant  
523           Coffee Shop         Monument / Landmark     Italian Restaurant  
526              Bus Stop            Sushi Restaurant     Turkish Restaurant  
527           Golf Course  Construction & Landscaping                   Park  
529                  Park          Italian Restaurant      French Restaurant  

[293 rows x 12 columns]

Cluster 5

In [52]:
london_data_nonan.loc[london_data_nonan['Cluster Labels'] == 5, london_data_nonan.columns[[1] + list(range(5, london_data_nonan.shape[1]))]]

town  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
121  LONDON               5  Gym / Fitness Center              Bus Stop   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
121        Clothing Store           Supermarket        Hardware Store   

    6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
121           Zoo Exhibit   Filipino Restaurant               Exhibit   

    9th Most Common Venue 10th Most Common Venue  
121    Falafel Restaurant         Farmers Market



---



# Exploring Paris

## Neighbourhoods of Paris

### Data Collection

We read the json data with pandas.

In [62]:
!wget -q -O 'france-data.json' https://www.data.gouv.fr/fr/datasets/r/e88c6fda-1d09-42a0-a069-606d3259114e
print("Data Downloaded!")

Data Downloaded!


In [61]:
paris_raw = pd.read_json('france-data.json')
paris_raw.head()

ValueError: Unexpected character found when decoding 'false'

### Data Preprocessing

We break down each of the nested fields and create the dataframe that we need

In [37]:
paris_field_data = pd.DataFrame()
for f in paris_raw.fields:
    dict_new = f
    paris_field_data = paris_field_data.append(dict_new, ignore_index=True)

paris_field_data.head()

code_arr code_cant code_comm  ...            statut superficie z_moyen
0        3        34       003  ...  Chef-lieu canton      232.0    70.0
1        1        09       430  ...    Commune simple      535.0   101.0
2        2        32       412  ...    Commune simple      252.0    62.0
3        2        24       598  ...  Chef-lieu canton      394.0    59.0
4        2        32       040  ...    Commune simple      711.0    69.0

[5 rows x 17 columns]

### Feature Selection

We take the columns that we require, in case of paris it would be the geo_point_2d, nom_dept, nom_comm and postal_code

In [38]:
df_2 = paris_field_data[['postal_code','nom_comm','nom_dept','geo_point_2d']]
df_2

postal_code  ...                             geo_point_2d
0          94110  ...   [48.80588035965699, 2.333510249842654]
1          77178  ...   [49.07375705850074, 2.787942211427862]
2          77930  ...   [48.47430724856196, 2.510781898359126]
3          95230  ...   [48.98840778751709, 2.300499783419528]
4          77310  ...   [48.51197791258124, 2.569919095331142]
...          ...  ...                                      ...
1295       77230  ...   [49.00336689019137, 2.739932368905603]
1296       91350  ...    [48.6568896115566, 2.387896478580817]
1297       77440  ...   [48.98705524224735, 3.074558248708759]
1298       77220  ...  [48.74953288926905, 2.7262357852056303]
1299       77580  ...   [48.84031218010945, 2.889716285735505]

[1300 rows x 4 columns]

We have managed to collect 1300 records.

### Feature Engineering

We localize to only include Paris

In [39]:
df_paris = df_2[df_2['nom_dept'].str.contains('PARIS')].reset_index(drop=True)
df_paris

postal_code  ...                             geo_point_2d
0        75010  ...   [48.87602855694339, 2.361112904561707]
1        75016  ...   [48.86039876035177, 2.262099559395783]
2        75009  ...   [48.87689616237872, 2.337460241388529]
3        75015  ...   [48.84015541860987, 2.293559372435076]
4        75002  ...   [48.86790337886785, 2.344107166658533]
5        75011  ...   [48.85941549762748, 2.378741060237548]
6        75005  ...  [48.844508659617546, 2.349859385560182]
7        75019  ...   [48.88686862295828, 2.384694327870042]
8        75020  ...   [48.86318677744551, 2.400819826729021]
9        75003  ...   [48.86305413181178, 2.359361058970589]
10       75006  ...   [48.84896809191946, 2.332670898588416]
11       75018  ...  [48.892735074561706, 2.348711933867703]
12       75008  ...   [48.87252726662346, 2.312582560420059]
13       75013  ...   [48.82871768452136, 2.362468228516128]
14       75012  ...   [48.83515623066034, 2.419807034965275]
15       75007  ...   [48.85608259819694, 2.312438687733857]
16       75001  ...    [48.8626304851685, 2.336293446550539]
17       75004  ...  [48.854228281954754, 2.357361938142205]
18       75017  ...   [48.88733716648682, 2.307485559493426]
19       75014  ...   [48.82899321160942, 2.327100883257538]

[20 rows x 4 columns]

In [40]:
df_paris.shape

(20, 4)

We have managed to bring down the records to just 20 from 1300. Paris data is quite small than initally thought

In [41]:
df_paris.dtypes

postal_code     object
nom_comm        object
nom_dept        object
geo_point_2d    object
dtype: object

## Gelocations of the Neighbourhoods of Paris


We don't need to get the geo coordinates using an external data source or collect it with the arcgis API call since we already have it stored in the *geo_point_2d* column as a tuple in the *df_paris* dataframe.

Checking one of the geo coordinates.

In [87]:
df_paris['geo_point_2d'][0]

[48.87602855694339, 2.361112904561707]

In [43]:
temp1 = df_paris['geo_point_2d']
temp1

0      [48.87602855694339, 2.361112904561707]
1      [48.86039876035177, 2.262099559395783]
2      [48.87689616237872, 2.337460241388529]
3      [48.84015541860987, 2.293559372435076]
4      [48.86790337886785, 2.344107166658533]
5      [48.85941549762748, 2.378741060237548]
6     [48.844508659617546, 2.349859385560182]
7      [48.88686862295828, 2.384694327870042]
8      [48.86318677744551, 2.400819826729021]
9      [48.86305413181178, 2.359361058970589]
10     [48.84896809191946, 2.332670898588416]
11    [48.892735074561706, 2.348711933867703]
12     [48.87252726662346, 2.312582560420059]
13     [48.82871768452136, 2.362468228516128]
14     [48.83515623066034, 2.419807034965275]
15     [48.85608259819694, 2.312438687733857]
16      [48.8626304851685, 2.336293446550539]
17    [48.854228281954754, 2.357361938142205]
18     [48.88733716648682, 2.307485559493426]
19     [48.82899321160942, 2.327100883257538]
Name: geo_point_2d, dtype: object

In [45]:
paris_latlng = df_paris['geo_point_2d'].astype('str')

Spliting the geo_point_2d column into latitude and longitude.

### Latitude

In [46]:
paris_lat = paris_latlng.apply(lambda x: x.split(',')[0])
paris_lat = paris_lat.apply(lambda x: x.lstrip('['))
paris_lat

0      48.87602855694339
1      48.86039876035177
2      48.87689616237872
3      48.84015541860987
4      48.86790337886785
5      48.85941549762748
6     48.844508659617546
7      48.88686862295828
8      48.86318677744551
9      48.86305413181178
10     48.84896809191946
11    48.892735074561706
12     48.87252726662346
13     48.82871768452136
14     48.83515623066034
15     48.85608259819694
16      48.8626304851685
17    48.854228281954754
18     48.88733716648682
19     48.82899321160942
Name: geo_point_2d, dtype: object

### Longitude

In [47]:
paris_lng = paris_latlng.apply(lambda x: x.split(',')[1])
paris_lng = paris_lng.apply(lambda x: x.rstrip(']'))
paris_lng

0      2.361112904561707
1      2.262099559395783
2      2.337460241388529
3      2.293559372435076
4      2.344107166658533
5      2.378741060237548
6      2.349859385560182
7      2.384694327870042
8      2.400819826729021
9      2.359361058970589
10     2.332670898588416
11     2.348711933867703
12     2.312582560420059
13     2.362468228516128
14     2.419807034965275
15     2.312438687733857
16     2.336293446550539
17     2.357361938142205
18     2.307485559493426
19     2.327100883257538
Name: geo_point_2d, dtype: object

In [48]:
paris_geo_lat  = pd.DataFrame(paris_lat.astype(float))
paris_geo_lat.columns=['Latitude']
paris_geo_lng = pd.DataFrame(paris_lng.astype(float))
paris_geo_lng.columns=['Longitude']

Preparing our combined data by dropping the geo_point_2d column from our previously stored df_paris and concatenating with the latitude and longitude extracted from it

In [49]:
paris_combined_data = pd.concat([df_paris.drop('geo_point_2d', axis=1), paris_geo_lat, paris_geo_lng], axis=1)
paris_combined_data

postal_code                  nom_comm nom_dept   Latitude  Longitude
0        75010  PARIS-10E-ARRONDISSEMENT    PARIS  48.876029   2.361113
1        75016  PARIS-16E-ARRONDISSEMENT    PARIS  48.860399   2.262100
2        75009   PARIS-9E-ARRONDISSEMENT    PARIS  48.876896   2.337460
3        75015  PARIS-15E-ARRONDISSEMENT    PARIS  48.840155   2.293559
4        75002   PARIS-2E-ARRONDISSEMENT    PARIS  48.867903   2.344107
5        75011  PARIS-11E-ARRONDISSEMENT    PARIS  48.859415   2.378741
6        75005   PARIS-5E-ARRONDISSEMENT    PARIS  48.844509   2.349859
7        75019  PARIS-19E-ARRONDISSEMENT    PARIS  48.886869   2.384694
8        75020  PARIS-20E-ARRONDISSEMENT    PARIS  48.863187   2.400820
9        75003   PARIS-3E-ARRONDISSEMENT    PARIS  48.863054   2.359361
10       75006   PARIS-6E-ARRONDISSEMENT    PARIS  48.848968   2.332671
11       75018  PARIS-18E-ARRONDISSEMENT    PARIS  48.892735   2.348712
12       75008   PARIS-8E-ARRONDISSEMENT    PARIS  48.872527   2.312583
13       75013  PARIS-13E-ARRONDISSEMENT    PARIS  48.828718   2.362468
14       75012  PARIS-12E-ARRONDISSEMENT    PARIS  48.835156   2.419807
15       75007   PARIS-7E-ARRONDISSEMENT    PARIS  48.856083   2.312439
16       75001  PARIS-1ER-ARRONDISSEMENT    PARIS  48.862630   2.336293
17       75004   PARIS-4E-ARRONDISSEMENT    PARIS  48.854228   2.357362
18       75017  PARIS-17E-ARRONDISSEMENT    PARIS  48.887337   2.307486
19       75014  PARIS-14E-ARRONDISSEMENT    PARIS  48.828993   2.327101

### Co-ordinates for Paris

In [50]:
paris = geocode(address='Paris, France, FR')[0]
paris_lng_coords = paris['location']['x']
paris_lat_coords = paris['location']['y']
print("The geolocation of Paris: ", paris_lat_coords, paris_lng_coords)

The geolocation of Paris:  48.85341000000005 2.3488000000000397


## Visualize the Map of Paris

In [52]:
# Creating the map of Paris
map_Paris= folium.Map(location=[paris_lat_coords, paris_lng_coords], zoom_start=12)
map_Paris

# adding markers to map
for latitude, longitude, borough, town in zip(paris_combined_data['Latitude'], paris_combined_data['Longitude'], paris_combined_data['nom_comm'], paris_combined_data['nom_dept']):
    label = '{}, {}'.format(town, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='Blue',
        fill=True,
        fill_opacity=0.8
        ).add_to(map_Paris)  
    
map_Paris

### Venues in Paris

Using our previously defined function. Let's get the neaby venues present in each neighbourhood of Paris

In [88]:
venues_in_Paris = getNearbyVenues(paris_combined_data['nom_comm'], paris_combined_data['Latitude'], paris_combined_data['Longitude'])

PARIS-10E-ARRONDISSEMENT
PARIS-16E-ARRONDISSEMENT
PARIS-9E-ARRONDISSEMENT
PARIS-15E-ARRONDISSEMENT
PARIS-2E-ARRONDISSEMENT
PARIS-11E-ARRONDISSEMENT
PARIS-5E-ARRONDISSEMENT
PARIS-19E-ARRONDISSEMENT
PARIS-20E-ARRONDISSEMENT
PARIS-3E-ARRONDISSEMENT
PARIS-6E-ARRONDISSEMENT
PARIS-18E-ARRONDISSEMENT
PARIS-8E-ARRONDISSEMENT
PARIS-13E-ARRONDISSEMENT
PARIS-12E-ARRONDISSEMENT
PARIS-7E-ARRONDISSEMENT
PARIS-1ER-ARRONDISSEMENT
PARIS-4E-ARRONDISSEMENT
PARIS-17E-ARRONDISSEMENT
PARIS-14E-ARRONDISSEMENT


Sampling our Data

In [89]:
venues_in_Paris.head()

Neighbourhood  ...            Venue Category
0  PARIS-10E-ARRONDISSEMENT  ...  Mediterranean Restaurant
1  PARIS-10E-ARRONDISSEMENT  ...         French Restaurant
2  PARIS-10E-ARRONDISSEMENT  ...                      Café
3  PARIS-10E-ARRONDISSEMENT  ...                      Café
4  PARIS-10E-ARRONDISSEMENT  ...         Food & Drink Shop

[5 rows x 5 columns]

In [90]:
venues_in_Paris.shape

(1324, 5)

We have managed to collect 1324 venue records for the neighbourhoods in Paris

### Grouping by Venue Categories
We need to now see how many Venue Categories are there for further processing

In [91]:
venues_in_Paris.groupby('Venue Category').max()

Neighbourhood  ...                                     Venue
Venue Category                                    ...                                          
Afghan Restaurant       PARIS-11E-ARRONDISSEMENT  ...                               Afghanistan
African Restaurant       PARIS-9E-ARRONDISSEMENT  ...                         Wally Le Saharien
American Restaurant     PARIS-19E-ARRONDISSEMENT  ...                                  Harper's
Antique Shop             PARIS-9E-ARRONDISSEMENT  ...                   Hôtel des Ventes Drouot
Argentinian Restaurant   PARIS-3E-ARRONDISSEMENT  ...                                     Anahi
...                                          ...  ...                                       ...
Wine Bar                 PARIS-9E-ARRONDISSEMENT  ...                          Vingt Vins d'Art
Wine Shop                PARIS-3E-ARRONDISSEMENT  ...                            Trois Fois Vin
Women's Store            PARIS-2E-ARRONDISSEMENT  ...                           & Other Stories
Zoo                     PARIS-12E-ARRONDISSEMENT  ...                  Parc zoologique de Paris
Zoo Exhibit             PARIS-12E-ARRONDISSEMENT  ...  Grande Serre du Parc Zoologique de Paris

[198 rows x 4 columns]

198 records in Paris. Paris is also multicultural and diverse as London

### One Hot Encoding 
We need to Encode our venue categories to get a better result for our clustering

In [92]:
Paris_venue_cat = pd.get_dummies(venues_in_Paris[['Venue Category']], prefix="", prefix_sep="")
Paris_venue_cat

Afghan Restaurant  African Restaurant  ...  Zoo  Zoo Exhibit
0                     0                   0  ...    0            0
1                     0                   0  ...    0            0
2                     0                   0  ...    0            0
3                     0                   0  ...    0            0
4                     0                   0  ...    0            0
...                 ...                 ...  ...  ...          ...
1319                  0                   0  ...    0            0
1320                  0                   0  ...    0            0
1321                  0                   0  ...    0            0
1322                  0                   0  ...    0            0
1323                  0                   0  ...    0            0

[1324 rows x 198 columns]

Adding Neighbourhoods to our Data

In [93]:
Paris_venue_cat['Neighbourhood'] = venues_in_Paris['Neighbourhood'] 

# moving neighborhood column to the first column
fixed_columns = [Paris_venue_cat.columns[-1]] + list(Paris_venue_cat.columns[:-1])
Paris_venue_cat = Paris_venue_cat[fixed_columns]

Paris_venue_cat.head()

Neighbourhood  Afghan Restaurant  ...  Zoo  Zoo Exhibit
0  PARIS-10E-ARRONDISSEMENT                  0  ...    0            0
1  PARIS-10E-ARRONDISSEMENT                  0  ...    0            0
2  PARIS-10E-ARRONDISSEMENT                  0  ...    0            0
3  PARIS-10E-ARRONDISSEMENT                  0  ...    0            0
4  PARIS-10E-ARRONDISSEMENT                  0  ...    0            0

[5 rows x 199 columns]

### Venue categories mean value
We will group the Neighbourhoods and calculate the mean venue categories value in each Neighbourhood

In [94]:
Paris_grouped = Paris_venue_cat.groupby('Neighbourhood').mean().reset_index()
Paris_grouped.head()

Neighbourhood  Afghan Restaurant  ...  Zoo  Zoo Exhibit
0  PARIS-10E-ARRONDISSEMENT           0.000000  ...  0.0          0.0
1  PARIS-11E-ARRONDISSEMENT           0.021739  ...  0.0          0.0
2  PARIS-12E-ARRONDISSEMENT           0.000000  ...  0.2          0.2
3  PARIS-13E-ARRONDISSEMENT           0.000000  ...  0.0          0.0
4  PARIS-14E-ARRONDISSEMENT           0.000000  ...  0.0          0.0

[5 rows x 199 columns]

### Top venue categories
 
Reusing our previously defined function to get the top venue categories in the neighbourhoods of Paris.

In [95]:
# create a new dataframe for Paris
neighborhoods_venues_sorted_paris = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted_paris['Neighbourhood'] = Paris_grouped['Neighbourhood']

for ind in np.arange(Paris_grouped.shape[0]):
    neighborhoods_venues_sorted_paris.iloc[ind, 1:] = return_most_common_venues(Paris_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted_paris.head()

Neighbourhood  ... 10th Most Common Venue
0  PARIS-10E-ARRONDISSEMENT  ...            Pizza Place
1  PARIS-11E-ARRONDISSEMENT  ...       Asian Restaurant
2  PARIS-12E-ARRONDISSEMENT  ...      Food & Drink Shop
3  PARIS-13E-ARRONDISSEMENT  ...           Dance Studio
4  PARIS-14E-ARRONDISSEMENT  ...                   Café

[5 rows x 11 columns]

## Model Building

### K Means
Let's cluster the city of Paris to roughly 5 to make it easier to analyze. 

We use the K Means clustering technique to do so.

In [96]:
# set number of clusters
k_num_clusters = 5

Paris_grouped_clustering = Paris_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans_Paris = KMeans(n_clusters=k_num_clusters, random_state=0).fit(Paris_grouped_clustering)
kmeans_Paris

KMeans(algorithm='auto', copy_x=True, init='k-means++', max_iter=300,
       n_clusters=5, n_init=10, n_jobs=None, precompute_distances='auto',
       random_state=0, tol=0.0001, verbose=0)

### Labelling Clustered Data

In [97]:
kmeans_Paris.labels_

array([4, 4, 0, 1, 2, 4, 3, 2, 4, 4, 4, 4, 4, 4, 4, 4, 4, 2, 2, 4], dtype=int32)

So our model has labeled the city, we insert it in our data.

In [98]:
neighborhoods_venues_sorted_paris.insert(0, 'Cluster Labels', kmeans_Paris.labels_ +1)

Join *paris_combined_data* with our neighbourhood venues sorted to add latitude & longitude for each of the neighborhood to prepare it for plotting

In [99]:
paris_data = paris_combined_data

paris_data = paris_data.join(neighborhoods_venues_sorted_paris.set_index('Neighbourhood'), on='nom_comm')

paris_data.head()

postal_code  ... 10th Most Common Venue
0       75010  ...            Pizza Place
1       75016  ...                   Pool
2       75009  ...            Pizza Place
3       75015  ...            Coffee Shop
4       75002  ...                 Bistro

[5 rows x 16 columns]


Drop all the NaN values to prevent data skew

In [100]:
paris_data_nonan = paris_data.dropna(subset=['Cluster Labels'])

### Visualizing the clustered neighbourhood
Let's plot the clusters

In [101]:
map_clusters_paris = folium.Map(location=[paris_lat_coords, paris_lng_coords], zoom_start=12)

# set color scheme for the clusters
x = np.arange(k_num_clusters)
ys = [i + x + (i*x)**2 for i in range(k_num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(paris_data_nonan['Latitude'], paris_data_nonan['Longitude'], paris_data_nonan['nom_comm'], paris_data_nonan['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + ' ' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)],
        fill_opacity=0.8
        ).add_to(map_clusters_paris)
        
map_clusters_paris

## Examining our Clusters

Cluster 1

In [102]:
paris_data_nonan.loc[paris_data_nonan['Cluster Labels'] == 1, paris_data_nonan.columns[[1] + list(range(5, paris_data_nonan.shape[1]))]]

nom_comm  ...  10th Most Common Venue
14  PARIS-12E-ARRONDISSEMENT  ...       Food & Drink Shop

[1 rows x 12 columns]

Cluster 2

In [103]:
paris_data_nonan.loc[paris_data_nonan['Cluster Labels'] == 2, paris_data_nonan.columns[[1] + list(range(5, paris_data_nonan.shape[1]))]]

nom_comm  ...  10th Most Common Venue
13  PARIS-13E-ARRONDISSEMENT  ...            Dance Studio

[1 rows x 12 columns]

Cluster 3

In [104]:
paris_data_nonan.loc[paris_data_nonan['Cluster Labels'] == 3, paris_data_nonan.columns[[1] + list(range(5, paris_data_nonan.shape[1]))]]

nom_comm  ...     10th Most Common Venue
12   PARIS-8E-ARRONDISSEMENT  ...  Middle Eastern Restaurant
15   PARIS-7E-ARRONDISSEMENT  ...             Ice Cream Shop
18  PARIS-17E-ARRONDISSEMENT  ...             Breakfast Spot
19  PARIS-14E-ARRONDISSEMENT  ...                       Café

[4 rows x 12 columns]

Cluster 4

In [105]:
paris_data_nonan.loc[paris_data_nonan['Cluster Labels'] == 4, paris_data_nonan.columns[[1] + list(range(5, paris_data_nonan.shape[1]))]]

nom_comm  ...  10th Most Common Venue
1  PARIS-16E-ARRONDISSEMENT  ...                    Pool

[1 rows x 12 columns]

Cluster 5

In [106]:
paris_data_nonan.loc[paris_data_nonan['Cluster Labels'] == 5, paris_data_nonan.columns[[1] + list(range(5, paris_data_nonan.shape[1]))]]

nom_comm  ...  10th Most Common Venue
0   PARIS-10E-ARRONDISSEMENT  ...             Pizza Place
2    PARIS-9E-ARRONDISSEMENT  ...             Pizza Place
3   PARIS-15E-ARRONDISSEMENT  ...             Coffee Shop
4    PARIS-2E-ARRONDISSEMENT  ...                  Bistro
5   PARIS-11E-ARRONDISSEMENT  ...        Asian Restaurant
6    PARIS-5E-ARRONDISSEMENT  ...                  Bistro
7   PARIS-19E-ARRONDISSEMENT  ...     Japanese Restaurant
8   PARIS-20E-ARRONDISSEMENT  ...                   Hotel
9    PARIS-3E-ARRONDISSEMENT  ...      Chinese Restaurant
10   PARIS-6E-ARRONDISSEMENT  ...                Fountain
11  PARIS-18E-ARRONDISSEMENT  ...                   Plaza
16  PARIS-1ER-ARRONDISSEMENT  ...              Art Museum
17   PARIS-4E-ARRONDISSEMENT  ...                  Bistro

[13 rows x 12 columns]



---




# Results and Discussion

The neighbourhoods of London are very mulitcultural. There are a lot of different cusines including Indian, Italian, Turkish and Chinese. London seems to take a step further in this direction by having a lot of Restaurants, bars, juice bars, coffee shops, Fish and Chips shop and Breakfast spots. It has a lot of shopping options too with that of the Flea markets, flower shops, fish markets, Fishing stores, clothing stores. The main modes of transport seem to be Buses and trains. For leisure, the neighbourhoods are set up to have lots of parks, golf courses, zoo, gyms and Historic sites.

Overall, the city of London offers a multicultural, diverse and certainly an entertaining experience.

Paris is relatively small in size geographically. It has a wide variety of cusines and eateries including French, Thai, Cambodian, Asian, Chinese etc. There are a lot of hangout spots including many Restaurants and Bars. Paris has a lot of Bistro's. Different means of public transport in Paris which includes buses, bikes, boats or ferries. For leisure and sight seeing, there are a lot of Plazas, Trails, Parks, Historic sites, clothing shops, Art galleries and Museums. Overall, Paris seems like the relaxing vacation spot with a mix of lakes, historic spots and a wide variety of cusines to try out.


# Conclusion

The purpose of this project was to explore the cities of London and Paris and see how attractive it is to potential tourists and migrants. We explored both the cities based on their postal codes and then extrapolated the common venues present in each of the neighbourhoods finally concluding with clustering similar neighbourhoods together.

We could see that each of the neighbourhoods in both the cities have a wide variety of experiences to offer which is unique in it's own way. The cultural diversity is quite evident which also gives the feeling of a sense of inclusion.

Both Paris and London seem to offer a vacation stay or a romantic gateaway with a lot of places to explore, beautiful landscapes and a wide variety of culture.Overall, it's upto the stakeholders to decide which experience they would prefer more and which would more to their liking.